ffi/derive/src/export.rs

```diff
@@ -0,0 +1,276 @@
+    let mut stmts = quote! {};
+
+    if let Some(arg) = &fn_descriptor.receiver {
+        let arg_name = &arg.name();
+
+        stmts = if matches!(arg.src_type(), Type::Path(_)) {
+            quote! {let __tmp_handle = #arg_name.read();}
+        } else {
+            gen_arg_ffi_to_src(arg, false)
+        };
+    }
+
+    for arg in &fn_descriptor.input_args {
+        stmts.extend(gen_arg_ffi_to_src(arg, false));
+    }
+
+    stmts
+}
+
+fn gen_method_call_stmt(fn_descriptor: &FnDescriptor) -> TokenStream {
+    let method_name = fn_descriptor.method_name;
+    let self_type = fn_descriptor.self_ty;
+
+    let receiver = fn_descriptor.receiver.as_ref();
+    let self_arg_name = receiver.map_or_else(Vec::new, |arg| {
+        if matches!(arg.src_type(), Type::Path(_)) {
+            return vec![Ident::new("__tmp_handle", Span::call_site())];
+        }
+
+        vec![arg.name().clone()]
+    });
+
+    let fn_arg_names = fn_descriptor.input_args.iter().map(Arg::name);
+    let method_call = quote! {#self_type::#method_name(#(#self_arg_name,)* #(#fn_arg_names),*)};
+
+    fn_descriptor.output_arg.as_ref().map_or_else(
+        || quote! {#method_call;},
+        |output_arg| {
+            let output_arg_name = &output_arg.name();
+
+            quote! {
+                let __out_ptr = #output_arg_name;
+                let #output_arg_name = #method_call;
+            }
+        },
+    )
+}
+
+fn ffi_output_arg<'tmp: 'ast, 'ast>(
+    fn_descriptor: &'tmp FnDescriptor<'ast>,
+) -> Option<&'ast crate::impl_visitor::ReturnArg<'ast>> {
+    fn_descriptor.output_arg.as_ref().and_then(|output_arg| {
+        if let Some(receiver) = &fn_descriptor.receiver {
+            if receiver.name() == output_arg.name() {
+                return None;
+            }
+        }
+
+        Some(output_arg)
+    })
+}
+
+pub fn gen_arg_ffi_to_src(arg: &impl crate::impl_visitor::Arg, is_output: bool) -> TokenStream {
+    let (arg_name, src_type) = (arg.name(), arg.src_type_resolved());
+
+    if is_output {
+        let mut stmt = quote! {
+            let mut store = ();
+            let #arg_name: #src_type = iroha_ffi::TryFromReprC::try_from_repr_c(#arg_name, &mut store)?;
+        };
+
+        if let Type::Reference(ref_type) = &src_type {
+            let elem = &ref_type.elem;
+
+            stmt.extend(if ref_type.mutability.is_some() {
+                quote! {
+                    // NOTE: Type having `type TryFromReprC::Store = ()` will never reference
+                    // local context, i.e. it's lifetime can be attached to that of the wrapping fn
+                    unsafe { &mut *(#arg_name as *mut #elem) }
+                }
+            } else {
+                quote! {
+                    unsafe { &*(#arg_name as *const #elem) }
+                }
+            });
+        }
+
+        return stmt;
+    }
+
+    quote! {
+        let mut store = Default::default();
```

Comment:
Since proc-macros are unhygienic, consider using fully quilified paths like `core::default::Default`, to avoid possibility of using different `Default` defined for crate.

Reply:
yeah, I tried to do that mostly, sometimes I missed it. These will be found if conflicting with something else